File: python/mirage/commands/builtin/s3/wc/wc.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from collections.abc import AsyncIterator

from mirage.accessor.s3 import S3Accessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.wc import WCCounts, format_wc
from mirage.commands.builtin.generic.wc import wc as generic_wc
from mirage.commands.builtin.s3._provision import file_read_provision
from mirage.commands.builtin.utils.stream import _resolve_source
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.s3.glob import resolve_glob
from mirage.core.s3.read import read_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("wc", resource="s3", spec=SPECS["wc"], provision=file_read_provision)
async def wc(
    accessor: S3Accessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    args_l: bool = False,
    w: bool = False,
    c: bool = False,
    m: bool = False,
    L: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if paths:
        paths = await resolve_glob(accessor, paths, index)
        outputs: list[str] = []
        totals = WCCounts()
        for p in paths:
            data = await read_bytes(accessor, p)
            counts = await generic_wc(data)
            outputs.append(
                format_wc(counts,
                          args_l=args_l,
                          w=w,
                          c=c,
                          m=m,
                          L=L,
                          label=p.original))
            totals.merge(counts)
        if len(paths) > 1:
            outputs.append(
                format_wc(totals,
                          args_l=args_l,
                          w=w,
                          c=c,
                          m=m,
                          L=L,
                          label="total"))
        return "\n".join(outputs).encode(), IOResult()
    source = _resolve_source(stdin, "wc: missing operand")
    counts = await generic_wc(source)
    return format_wc(counts, args_l=args_l, w=w, c=c, m=m,
                     L=L).encode(), IOResult()

File: typescript/packages/core/src/commands/builtin/postgres/wc.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { PostgresAccessor } from '../../../accessor/postgres.ts'
import { countRows } from '../../../core/postgres/_client.ts'
import { resolveGlob } from '../../../core/postgres/glob.ts'
import { readStream } from '../../../core/postgres/read.ts'
import { detectScope } from '../../../core/postgres/scope.ts'
import { type ByteSource, IOResult } from '../../../io/types.ts'
import { type PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { formatRecords } from '../utils/output.ts'
import { formatWcLines, wcGeneric, type WcRow } from '../generic/wc.ts'
import { fileReadProvision } from './_provision.ts'

function rowsScope(p: PathSpec): { schema: string; entity: string } | null {
  const scope = detectScope(p)
  if (scope.level === 'entity_rows') {
    return { schema: scope.schema, entity: scope.entity }
  }
  return null
}

async function wcCommand(
  accessor: PostgresAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const f = opts.flags
  const resolved =
    paths.length > 0 ? await resolveGlob(accessor, paths, opts.index ?? undefined) : []
  // Line counts on tables/views come from a server-side COUNT(*) instead of
  // reading every row. -l only (default prints words and bytes too, which
  // needs the content).
  const countOnly =
    f.args_l === true && f.w !== true && f.c !== true && f.m !== true && f.L !== true
  if (countOnly && resolved.length > 0 && resolved.every((p) => rowsScope(p) !== null)) {
    const rows: WcRow[] = []
    let total = 0
    for (const p of resolved) {
      const scope = rowsScope(p)
      if (scope === null) continue
      const count = await countRows(accessor, scope.schema, scope.entity)
      rows.push({ values: [count], label: p.original })
      total += count
    }
    if (resolved.length > 1) rows.push({ values: [total], label: 'total' })
    const out: ByteSource = formatRecords(formatWcLines(rows))
    return [out, new IOResult()]
  }
  return wcGeneric(resolved, texts, opts, (p) => readStream(accessor, p, opts.index ?? undefined))
}

export const POSTGRES_WC = command({
  name: 'wc',
  resource: ResourceName.POSTGRES,
  spec: specOf('wc'),
  fn: wcCommand,
  provision: fileReadProvision,
})
